File: java/01-arrays-hashing/004-group-anagrams.java
/*
 * NeetCode 150 - Problem 4: Group Anagrams
 *
 * Approach 1:
 * Sort each word and use the sorted word as key.
 * Time Complexity: O(m * n log n)
 *
 * Approach 2 (Optimal):
 * Use a 26-length frequency count array for each word.
 * Convert the count array to a String key and group words
 * with the same frequency pattern using a HashMap.
 *
 * Time Complexity: O(m * n)
 * Space Complexity: O(m * n)
 */


import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {

        Map<String, List<String>> wordMap = new HashMap<>();

        for (String word : strs) {
            int[] count = new int[26];

            for (char ch : word.toCharArray()) {
                count[ch - 'a']++;
            }

            String key = Arrays.toString(count);

            wordMap.putIfAbsent(key, new ArrayList<>());
            wordMap.get(key).add(word);
        }

        return new ArrayList<>(wordMap.values());
    }
}

File: python/01-arrays-hashing/004-group-anagrams.py
"""
NeetCode 150 - Problem 4: Group Anagrams

Approach:
Use a dictionary to group words by their character frequency count.
For each word, create a 26-length frequency array (for 'a' to 'z').
Convert this array into a tuple (since lists cannot be dictionary keys)
and use it as the key to group anagrams together.

Time Complexity: O(m * n)
    m = number of strings
    n = average length of each string

Space Complexity: O(m * n)
"""

from typing import List
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = defaultdict(list)

        for word in strs:
            count = [0] * 26  # Frequency count for 'a' to 'z'

            for ch in word:
                count[ord(ch) - ord('a')] += 1

            res[tuple(count)].append(word)

        return list(res.values())
